### PR TITLE
Add OCAP AAR Functionality

### PR DESCRIPTION
**Pull request by TyroneMF**
Mission Type
- `missionType` config value added to `description.ext` under the others that get filled out will suffice.

Chat commands:
- `#tac-capture` calls server event to begin recording.
- `#tac-save` calls server event to save the replay.

Does not support multiple recordings in the same mission (OCAP limitation somewhere), Events added to `admin\initServer.sqf`

Automation:
- `tac_getWeekDay` is now part of the time until start. Saturday = 6, Sunday = 0
- Automatic start will begin at 14:00Z on Saturdays and 15:00Z on any other day.
- If time has already passed 14:00Z or 15:00Z it will begin recording immediately.
- All missions except Gimmicks are recorded. (reason below)
- PerFrameHandler inside waitAndExecute to check `ocap_capture` and players are above 5 to start recording
- if `ocap_capture` is true and there are less than 5 players it automatically ends.
- Recording is saved with the correct "tag" grabbed from `tac_type`
- If mission is switched via `#missions` then `MpEnded` eventhandler fires with end capture event.
- If player count drops below 5 it will save the capture, if it raises back above 5 it will try to restart it (currently not possible)

Mission Name:
- Mission Name is auto called using `_timeUntilStart + 60` to cover if it's an instant start.

~~Tested and working.~~

### CODE REVIEW

**Comment by jonpas** [comment]
With https://github.com/OCAP2/addon/pull/27 we should do:
- Mission Type variable in template, can be `Contract`, `Non-Contract`, `Training`, `Special`, `Special (PvP)`, `Gimmick`.
- Use Mission Type in `ocap_fnc_exportData` to set the tag for filtering.
- Use Mission Type to determine record start (if `Contract` or `Non-Contract`):
  - Start at 1400z if Saturday, 1500z if Sunday (we start them 1h earlier for preparation).
  - Start immediately if already past 1400z/1500z.
  - Retain min. player count (5?) - if already 1400z/1500z, wait for player count either way

This should catch all our cases:
- `Contracts` on Saturdays launch at 1300z, start at 1400z.
- `Non-Contracts` on Sundays launch at 1400z, start at 1500z.
- All others launch on start time.

There is probably some other edge case we are not considering.

**Comment by TyroneMF** [comment]
> All others launch on start time.

No point in recording Gimmicks, most are custom built so it'll just look like we're running around on an empty map/area.

**Comment by TyroneMF** [comment]
Flags can just simply be added alongside regular `description.ext` values.

```
// Mission settings - title, description, author etc.
OnLoadName = "MISSION NAME";
OnLoadMission = "MISSION SUMMARY";
author = "YOUR NAME, Theseus Inc.";
missionType = "Contract"; // Values are "Contract", "Non-Contract", "Training", "Special", "Special (PvP), "PvP" or "Gimmick".
```

`missionType` doesn't exist and it'll never get added so i see no issues in using it.

File: template/admin/init.sqf [hunk 1]
```diff
@@ -15,3 +15,6 @@
     [_x, "theseus", 15, "label", "HELIOS CMND"] call acre_api_fnc_setPresetChannelField;
     [_x, "theseus"] call acre_api_fnc_setPreset;
 } forEach ["ACRE_PRC152", "ACRE_PRC117F"];
+
+// AAR Chat Commands
+[] call FUNC(chatCommands);
```
Comment (jonpas) [inline template/admin/init.sqf:20]: Either change the comment, or rename the function, it's AAR-only now.

File: template/description.ext [hunk 2]
```diff
@@ -4,6 +4,7 @@
 OnLoadName = "MISSION NAME";
 OnLoadMission = "MISSION SUMMARY";
 author = "YOUR NAME, Theseus Inc.";
+missionType = "Contract"; // Values are "Contract", "Non-Contract", "Training", "Special", "Special (PvP)", "PvP" or "Gimmick".
```
[the remark this answers is not in the document]
Comment (jonpas) [reply template/description.ext:7]: And have you verified you can _actually read_ this?

[the remark this answers is not in the document]
Comment (TyroneMF) [reply template/description.ext:7]: `getMissionConfigValue ["tac_type", ""];`
Showed `"Contract"` on my test.

File: template/admin/initServer.sqf [hunk 3]
```diff
@@ -1,13 +1,73 @@
...
+    //diag_log format ["[TAC] Mission Type String: %1", _missionTypeString];
+    ["", "", _missionTypeString] call ocap_fnc_exportData;
+};
+
+// Handle mission ending
+addMissionEventHandler ["MpEnded", _endAAR];
+
+if (_missionType != 5) then { // Handle Contract/Non-Contracts
+    [{
+        params ["_endAAR"];
```
Comment (jonpas) [inline template/admin/initServer.sqf:41]:
```suggestion
```

File: template/admin/initServer.sqf [hunk 4]
```diff
@@ -1,13 +1,73 @@
...
+        [{
+            params ["_args", "_handle"];
+            _args params ["_endAAR"];
```
Comment (jonpas) [inline template/admin/initServer.sqf:44]:
```suggestion
            params ["_endAAR", "_handle"];
```

File: template/admin/initServer.sqf [hunk 5]
```diff
@@ -1,13 +1,73 @@
...
+            private _playerCount = call CBA_fnc_players;
```
Comment (jonpas) [inline template/admin/initServer.sqf:45]:
```suggestion
            private _playerCount = count (call CBA_fnc_players);
```

File: template/admin/initServer.sqf [hunk 6]
```diff
@@ -1,13 +1,73 @@
...
+
+            if (!ocap_capture && {_playerCount >= 5}) then {
+                [] call ocap_fnc_init;
+            };
+
+            if (ocap_capture && {_playerCount < 5}) then {
+                call _endAAR
+            };
+        }, 60, [_endAAR]] call CBA_fnc_addPerFrameHandler;
+    }, [_endAAR], _timeUntilStart] call CBA_fnc_waitAndExecute;
```
Comment (jonpas) [inline template/admin/initServer.sqf:55]:
```suggestion
        }, 60, _this] call CBA_fnc_addPerFrameHandler;
    }, _endAAR, _timeUntilStart] call CBA_fnc_waitAndExecute;
```

File: template/admin/initServer.sqf [hunk 7]
```diff
@@ -1,13 +1,73 @@
...
+};
+
+// AAR Events for chat commands
+[QGVAR(captureAAR), {
+    [] call ocap_fnc_init;
```
Comment (jonpas) [inline template/admin/initServer.sqf:60]: Check `ocap_capture` here too.

File: template/admin/initServer.sqf [hunk 8]
```diff
@@ -1,13 +1,73 @@
...
+}] call CBA_fnc_addEventHandler;
+
+[QGVAR(saveAAR), {
+    [] call ocap_fnc_exportData;
```
Comment (jonpas) [inline template/admin/initServer.sqf:64]: Check `ocap_capture` here too.

File: template/functions/fn_chatCommands.sqf [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Tyrone
+ * Registers chat commands for use.
+ * call from init.sqf
```
Comment (jonpas) [inline template/functions/fn_chatCommands.sqf:5]:
```suggestion
 * Call from init.sqf.
```

**Review by jonpas — CHANGES_REQUESTED** [review]

File: template/admin/initServer.sqf [hunk 10]
```diff
@@ -1,13 +1,81 @@
...
+private _endAAR = {
+    private _missionType = getMissionConfigValue ["tac_type", ""];
+    private _missionTypeString = ["Contract", "Non-Contract", "Training", "Special", "PvP"] select _missionType;
+    //diag_log format ["[TAC] Mission Type: %1", _missionType];
+    //diag_log format ["[TAC] Mission Type String: %1", _missionTypeString];
+    [west, "", _missionTypeString] call ocap_fnc_exportData;
+};
+
+// Handle mission ending
+addMissionEventHandler ["MpEnded", _endAAR];
```
Comment (jonpas) [inline template/admin/initServer.sqf:40]:
```suggestion
addMissionEventHandler ["MPEnded", _endAAR];
```

File: template/admin/initServer.sqf [hunk 11]
```diff
@@ -1,13 +1,81 @@
...
+
+private _missionType = getMissionConfigValue ["tac_type", ""];
+
+ // Handle Contract/Non-Contracts
+if (_missionType != 5) then {
```
Comment (jonpas) [inline template/admin/initServer.sqf:45]:
```suggestion
// Only record Contract, Non-Contract, Training, Special, PvP
private _missionType = getMissionConfigValue ["tac_type", -1];
if (_missionType != -1 && {_missionType <= 4}) then {
```

File: template/admin/initServer.sqf [hunk 12]
```diff
@@ -1,13 +1,81 @@
...
+
+//diag_log format ["[TAC] Time Until Start: %1", _timeUntilStart];
+
+// Mission end function
+private _endAAR = {
+    private _missionType = getMissionConfigValue ["tac_type", ""];
+    private _missionTypeString = ["Contract", "Non-Contract", "Training", "Special", "PvP"] select _missionType;
+    //diag_log format ["[TAC] Mission Type: %1", _missionType];
+    //diag_log format ["[TAC] Mission Type String: %1", _missionTypeString];
+    [west, "", _missionTypeString] call ocap_fnc_exportData;
```
Comment (jonpas) [inline template/admin/initServer.sqf:36]:
```suggestion
    private _missionType = getMissionConfigValue ["tac_type", -1];
    if (_missionType != -1 && {_missionType <= 4}) then {
        private _missionTypeString = ["Contract", "Non-Contract", "Training", "Special", "PvP"] select _missionType;
        [sideAmbientLife, "", _missionTypeString] call ocap_fnc_exportData; // side must be given
    };
```

File: template/admin/initServer.sqf [hunk 13]
```diff
@@ -1,13 +1,81 @@
...
+
 // Mathematical Heresy
 systemTimeUTC params ["", "", "", "_hour", "_minute"];
-private _timeUntilStart = ((14 * 60) - (_hour * 60 + _minute)) * 60; // start time (1400z) - current time = time until start time
+private _weekDay = [systemTimeUTC] call tac_getWeekDay;
+private _startTime = [15, 14] select (_weekDay == 5); // 1400z on Saturday, 1500z otherwise
+private _timeUntilStart = ((_startTime * 60) - (_hour * 60 + _minute)) * 60; // start time - current time = time until start time
+_timeUntilStart = _timeUntilStart max 0;
+
+//diag_log format ["[TAC] Time Until Start: %1", _timeUntilStart];
```
Comment (jonpas) [inline template/admin/initServer.sqf:28]:
```suggestion
LOG_1("Time Until Start: %1",_timeUntilStart);
```
Check how this actually logs in RPT.

**Review by jonpas — CHANGES_REQUESTED** [review]

File: template/admin/initServer.sqf (hunk 13 above)
Comment (TyroneMF) [reply template/admin/initServer.sqf]: "For Later" - Doesn't show in RPT at all.

**Review by jonpas — CHANGES_REQUESTED** [review]

**Pull request merged**